[AIRFLOW-5088][AIP-24] Persisting serialized DAG in DB for webserver scalability

coufon opened the pull request:
Co-Authored-By: Kaxil Naik <<kaxilnaik@gmail.com>>
Co-Authored-By: Ash Berlin-Taylor <<ash_github@firemirror.com>>


Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira 5088](https://issues.apache.org/jira/browse/AIRFLOW-5088) issues and references them in the PR title.
  - [AIP-24](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-24+DAG+Persistence+in+DB+using+JSON+for+Airflow+Webserver+and+%28optional%29+Scheduler)

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

 - Based on https://github.com/apache/airflow/pull/5701, this PR implements functionalities including writing serialized DAGs to DB in scheduler, reading DAGs from DB in webserver, controlled by [core] dagcached
 - The goal is to decouple webserver from the DAG folder, instead it reads everything from database.
 - Rendering template by functions is an exception, in that case it needs to re-import DAG, because functions are stringified in serialized DAG.


### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
 - Added unit tests for airflow/models/serialized_dag.py
 - Tested basic UI functionanilties end-to-end

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

kaxil commented on airflow/models/dag.py:
```diff
@@ -52,6 +53,8 @@
 from airflow.utils.sqlalchemy import UtcDateTime, Interval
 from airflow.utils.state import State
 
+DAGCACHED_ENABLED = configuration.getboolean('core', 'dagcached', fallback=True)
```
We should probably just have this setting at one place (May be in `settings.py`) and import it from there.

kaxil replied:
WDYT?

coufon replied:
Good point. I will change this

coufon replied:
Done.

Commit:
[AIRFLOW-5088] persisting serialized DAG in DB for webserver scalability

milton0825 commented on airflow/models/dagbag.py:
```diff
@@ -416,6 +427,19 @@ def collect_dags(
+        # from the table by the scheduler job.
+        self.log.info("Filling up the DagBag from database")
+        self.dags = SerializedDagModel.read_all_dags()
+        Stats.gauge(
```
Can we use `Stats.timing` here? And you can just pass `timedelta` in.

coufon replied:
> Can we use `Stats.timing` here? And you can just pass `timedelta` in.

Good point! Done.

milton0825 commented on airflow/migrations/versions/d38e04c12aa2_add_serialized_dag_table.py:
```diff
@@ -0,0 +1,50 @@
+
+
+def upgrade():
+    op.create_table('serialized_dag',
```
Any reason why we need a new table instead of extending the existing `dag` table?

coufon replied:
Thanks for asking. There are a few benefits to create a few table:

- As this feature is new and may be unstable, we would like to minimize its influence on current Airflow code (and db performance), especially things on the critical path. Using a separate table is safer. At least we know the dag table's behavior is not changed
- Merging these two tables make the dag table wide: it needs a hash of file path, a 'DELETED' maker, and a JSON data column.
- Logically serialized_dag is a snapshot of the files we have in DAG folder, but the dag table is not

I am open for changes. Any ideas?

kaxil replied:
And as we want this feature to be completely optional till 2.0 - it is better to have a new table instead.

milton0825 replied:
Make sense

milton0825 commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,143 @@
+                    timezone.utcnow() - result.last_updated).total_seconds() < min_update_interval:
+                return
+        session.merge(cls(dag))
+        session.commit()
```
Should we do a `session.rollback()` in cased of encountering an exception?

milton0825 replied:
A handy func you can use is:
https://github.com/apache/airflow/blob/d5ad0761fd0b33cb89258ff6924c608c3e086680/airflow/utils/db.py#L33-L45

coufon replied:
Yes I think it is a best practice. I will add that.

kaxil replied:
@coufon - lets use that `create_session()` function instead of adding rolling back. May be you agreed to that only but just in-case

coufon replied:
Done.

Commit:
clean flake8 and pylint warnings

Commit:
use create_session instead of provide_session

Commit:
make dagbag.get_dag/dag.get_bag called by webserver/scheduler controlled by dagcached_enabled and only_from_file

Commit:
fix a degradation that inspect.getsource does not support functools.partial

Commit:
add unit tests of SerializedDagModel

kaxil commented on airflow/www/utils.py:
```diff
@@ -374,6 +375,38 @@ def get_chart_height(dag):
     return 600 + len(dag.tasks) * 10
 
 
+def get_python_source(x: Any) -> str:
```
```suggestion
def get_python_source(x: Any) -> Optional[str]:
```

if  :

```
if x is None:
        return None
```

coufon replied:
Yes. Already changed!

Commit:
fall back to loading from files when serialized DAGs are unavailable

kaxil commented on tests/dags/test_dag_serialization.py:
```diff
@@ -244,7 +244,6 @@ def test_serialization(self):
             # Verify JSON schema of serialized operators.
```
We should remove the hardcoding of dag names on the line 36

https://github.com/apache/airflow/pull/5743/files#diff-73d9177194b24994293316f69a960ec1R36

WDYT?
There is the below code we can use:

https://github.com/apache/airflow/blob/f6356f2cbbaeea9db2a98cce2c103d21c2651e62/tests/cli/test_cli.py#L112-L120

coufon replied:
I can work on this one.

coufon replied:
Done

kaxil committed:
Fix param names & remove unused params

kaxil commented on airflow/models/dagbag.py:
```diff
@@ -352,7 +376,8 @@ def collect_dags(
             include_examples=configuration.conf.getboolean('core', 'LOAD_EXAMPLES'),
-            safe_mode=configuration.conf.getboolean('core', 'DAG_DISCOVERY_SAFE_MODE')):
+            safe_mode=configuration.conf.getboolean('core', 'DAG_DISCOVERY_SAFE_MODE'),
+            dagcached_enabled=False):
```
We should remove this. It isn't used in this function.

L393 uses: `self.dagcached_enabled`

coufon replied:
Yes!

kaxil commented on airflow/settings.py:
```diff
@@ -310,3 +310,12 @@ def initialize():
+
+# If dagcached is enabled, scheduler writes serialized DAGs to DB, and webserevr
+# reads DAGs from DB instead of importing from files.
+DAGCACHED_ENABLED = conf.getboolean('core', 'dagcached', fallback=True)
```
The default behavior should be still parsing from files until 2.0

Changing it to the following

```suggestion
DAGCACHED_ENABLED = conf.getboolean('core', 'dagcached', fallback=False)
```

coufon replied:
I turned it on for testing and forgot to set it back.. Thanks for fixing it

kaxil committed:
Turning Serialisation Off by default

Commit:
remove hardcoded example DAGs in serialization unit tests

kaxil committed:
Add logs

Commit:
move enum module to enums to avoid confusion

kaxil committed:
Add config to airflow.cfg file

kaxil commented on airflow/migrations/versions/d38e04c12aa2_add_serialized_dag_table.py:
```diff
@@ -0,0 +1,51 @@
+                    sa.Column('fileloc', sa.String(length=2000), nullable=False),
+                    sa.Column('fileloc_hash', sa.Integer(), nullable=False),
+                    sa.Column('data', sa.Text(), nullable=False),
+                    sa.Column('last_updated', sa.DateTime(), nullable=False),
```
We need to add Timezone support here.

![image](https://user-images.githubusercontent.com/8811558/62812450-aa68d500-bafd-11e9-93a7-0d6be6179973.png)

The time shown in this table is not UTC which **might** cause issues as all the other table have UTC

**dag_run** table:

![image](https://user-images.githubusercontent.com/8811558/62812510-f0be3400-bafd-11e9-8aaf-5098593d761f.png)


I will add that.

coufon replied:
OK. Thanks!!

kaxil committed:
Add timezone to migration script

kaxil commented on airflow/migrations/versions/d38e04c12aa2_add_serialized_dag_table.py:
```diff
@@ -45,6 +47,37 @@ def upgrade():
+            column_name="last_updated",
+            type_=sa.TIMESTAMP(timezone=True),
+        )
+
```
@coufon  - Should work now

![image](https://user-images.githubusercontent.com/8811558/62814112-9a092800-bb06-11e9-8c5a-73863c2b58ff.png)

Commit:
fix operator displying as SerializedBaseOperator on UI

Commit:
add subdags into dagbag when loading from database

Commit:
clean mypy and pylint

ashb commented on airflow/config_templates/default_airflow.cfg:
```diff
@@ -204,6 +204,10 @@ dag_discovery_safe_mode = True
 
+# Whether to serialises DAGs and persist them in DB.
+# If set to True, Webserver reads from DB instead of parsing DAG files
+# More details: https://github.com/apache/airflow/pull/5743
```
We'll need a more "permenant" link here (probably to a page in the docs somewhere) before merging this.

kaxil replied:
OK. We can create `.rst` from the AIP and add it over here.

Let's do this at the end - Not resolving this comment till then

ashb commented on airflow/dag/serialization/serialized_baseoperator.py:
```diff
@@ -45,6 +45,8 @@ def __init__(self, *args, **kwargs):
         self.ui_color = BaseOperator.ui_color
         self.ui_fgcolor = BaseOperator.ui_fgcolor
         self.template_fields = BaseOperator.template_fields
+        # Not None for SubDagOperator.
```
"Not" in comment doesn't make sense?

milton0825 replied:
Can we add validation here to ensure that this field is `None` if the the operator is not `SubDagOperator`?

kaxil replied:
Added in tests https://github.com/apache/airflow/pull/5743/commits/c68ee581e0534b91e41f5e696394a9b1d6e12baa

ashb commented on airflow/models/dag.py:
```diff
@@ -1509,8 +1518,19 @@ def get_last_dagrun(self, session=None, include_externally_triggered=False):
+        """
+        dag = DagBag(
+            dag_folder=self.fileloc, dagcached_enabled=dagcached_enabled).get_dag(self.dag_id)
+        if dagcached_enabled and dag is None:
```
Something about this method feels... off to me: it's a bit hard to follow the logic, and the use of recursion to re-load the DagBag without the cache enabled makes it harder to follow the logic.

kaxil replied:
>There may be a delay for scheduler to write serialized DAG into database, loads from file in this case.

I guess the idea is if for any reason (connectivity probably or some other DB issue) this method will load the DAG from file, hence it uses recursion to reload DagBag without cache_enabled.

ashb commented on airflow/models/dagbag.py:
```diff
@@ -416,6 +449,27 @@ def collect_dags(
+                subdags[subdag.dag_id] = subdag
+        self.dags.update(subdags)
+
+        Stats.timing('collect_dags', timezone.utcnow() - start_dttm)
```
I don't think we should be emitting this metric here -- it will get mixed in with the same metric from the scheduler and will have a different mean/std dev etc. This should be a different metric.

kaxil replied:
How about just changing the metric name?

ashb replied:
Sure - `collect_db_dags`?

kaxil replied:
Updated

ashb commented on airflow/models/dagbag.py:
```diff
@@ -107,28 +112,51 @@ def size(self):
+        # FIXME: this exception should be removed in future, then webserver can be
+        # decoupled from DAG files.
+        if self.dagcached_enabled and not from_file_only:
+            print(self.dags)
```
```suggestion
```

kaxil replied:
Removed

ashb commented on airflow/models/dagbag.py:
```diff
@@ -107,28 +112,51 @@ def size(self):
+        if self.dagcached_enabled and dag is not None:
+            from airflow.dag.serialization.serialized_dag import SerializedDAG
+            enforce_from_file = isinstance(dag, SerializedDAG)
+            self.log.info("Processing DAG file to render template.")
```
This log shouldn't be here (but in the caller) -- all we know in this function is that we've been told to go to disk.

kaxil replied:
Done

ashb commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,155 @@
+            to reduce DB write rate, there is a minimal interval of updating serialized DAGs.
+        [scheduler] dag_dir_list_interval = 300 (s):
+            interval of deleting serialized DAGs in DB when the files are deleted, suggest
+            to use a smaller interval such as 60
```
What does this look like when rendered?

kaxil replied:
Sorry didn't understand this one, what do you mean?

ashb commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,155 @@
+
+        self.dag_id = dag.dag_id
+        self.fileloc = dag.full_filepath
+        self.fileloc_hash = SerializedDagModel.dag_fileloc_hash(self.fileloc)
```
```suggestion
        self.fileloc_hash = self.dag_fileloc_hash(self.fileloc)
```

kaxil replied:
Updated

ashb commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,155 @@
+        # hashing is needed because the length of fileloc is 2000 as an Airflow convention,
+        # which is over the limit of indexing. If we can reduce the length of fileloc, then
+        # hashing is not needed.
+        return int(0xFFFF & int(
```
What is this magic constant and int conversion? It is fairly opaque so needs a comment explaining what and why we do this.

On python3 at least we can do `int.from_bytes(hashlib.sha1("abc".encode('utf-8')).digest()[-2:], byteorder='big', signed=False)` -- though also: we should probably use more than the last two bytes for this.

kaxil replied:
Will let @coufon answer this one.

coufon replied:
@ashb  @kaxil 

I agree to use int.from_bytes(hashlib.sha1("abc".encode('utf-8')).digest()[-2:], byteorder='big', signed=False). It is more clear!

dag_fileloc_hash is not required to be unique. To avoid hash collision, we need to verify the full fileloc anyway, after querying by fileloc. So there is no concern of using 2 bytes in regard of correctness. Do we have other concerns?

kaxil replied:
Are we happy to use the previous version for Py2 ? @ashb @coufon ?

ashb replied:
Sounds good, yes.

ashb commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,155 @@
+        :param dag: a DAG to be written into database
+        :param min_update_interval: minimal interval in seconds to update serialized DAG
+        """
+        with db.create_session() as session:
```
Please use `@provide_session` so that we can control the lifetime of sessions outside of helper method (one of my long-term bug bears with Airflow is how we handle transactions)

kaxil replied:
Updated

ashb commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,155 @@
+        :param min_update_interval: minimal interval in seconds to update serialized DAG
+        """
+        with db.create_session() as session:
+            if min_update_interval is not None:
```
If it is none we do nothing? That is unexpected. Given this method is called `write` I would have expected it to just do the writing. Perhaps `update_db`/`update_row`/`update` is a more descriptive name?

kaxil replied:
If it is `None`, we write it to DB without checking the duration since last updating the row.

ashb commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,155 @@
+                result = session.query(cls.last_updated).filter(
+                    cls.dag_id == dag.dag_id).first()
+                if result is not None and (
+                        timezone.utcnow() - result.last_updated).total_seconds() < min_update_interval:
```
We're loading the entire dag row and the potentially large json blob here. Instead we should 

1) move the date check into SQL, and
2) use `exists()`

kaxil replied:
Done.

ashb commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,155 @@
+        for dag_id, data in serialized_dags:
+            dag = Serialization.from_json(data)  # type: Any
+            # Sanity check.
+            if dag.dag_id == dag_id:
```
If we are checking this we should do _something_ when it's false.

kaxil replied:
I will add a warning note.

kaxil replied:
Done in https://github.com/apache/airflow/pull/5743/commits/a70027b5dcbab9a65224d62f818c20c6f5c9ad5c

ashb commented on airflow/utils/log/logging_mixin.py:
```diff
@@ -46,9 +46,10 @@ def logger(self):
+    def log(self) -> Logger:
         try:
-            return self._log
+            # FIXME: LoggingMixin should have a default _log field.
```
This comment doesn't make sense - the default is handed 2 lines down.

coufon replied:
The concern is that mypy can not infer the type of self._log if self._log is not set in `__init__`, so we need a 'ignore mypy' comment.

kaxil replied:
![image](https://user-images.githubusercontent.com/8811558/62885204-277b9080-bd30-11e9-9ae7-e6efc16bbc9e.png)

This is what @coufon means

ashb replied:
Ah, we should replace this with 

```python
    @cached_property
    def log(self):
        return logging.root.getChild(
            self.__class__.__module__ + '.' + self.__class__.__name__
        )
```

Though do we have to change this at all in this PR? It's a change that seems unrelated to the topic of the PR.

kaxil replied:
Nope, we can ignore it.

ashb commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,155 @@
+        self.dag_id = dag.dag_id
+        self.fileloc = dag.full_filepath
+        self.fileloc_hash = SerializedDagModel.dag_fileloc_hash(self.fileloc)
+        self.data = Serialization.to_json(dag)
```
Either here, or inside to_json we should ensure that the JSON blob is valid - I want to minimize the chance of writing "odd"/invalid data in to our DB.

ashb replied:
Somewhere here I also want to add `version` field in to the JSON we store so that we can change this over time without having to "detect" the version based off what the stored data looks like.

ashb replied:
(This "version" is not the version/history of the dag, but the version/format of the serialization)

kaxil replied:
> Either here, or inside to_json we should ensure that the JSON blob is valid - I want to minimize the chance of writing "odd"/invalid data in to our DB.

Done in https://github.com/apache/airflow/pull/5743/commits/977a2fe3fd244bc3f366a1228324f8b3c58f30ac . WDYT - is that OK?

ashb commented on airflow/www/views.py:
```diff
@@ -1328,7 +1331,7 @@ def set_duration(tid):
         return self.render_template(
             'airflow/tree.html',
-            operators=sorted({op.__class__ for op in dag.tasks}, key=lambda x: x.__name__),
+            operators=sorted({op for op in dag.tasks}, key=lambda x: x.task_type),
```
Do the templates need updating for this change? Previously they were passed classes, now they are passed... something else.

coufon replied:
Already updated graph.html and tree.html. Use `op.task_type` instead of `op.__name__`.

kaxil committed:
Replace create_session with provide_session

kaxil committed:
Move the date check into SQL

kaxil committed:
Fix logging error

kaxil committed:
Sync Database schema and SQLAlchemy model

ashb commented on airflow/migrations/versions/d38e04c12aa2_add_serialized_dag_table.py:
```diff
@@ -61,6 +61,7 @@ def upgrade():
             table_name="serialized_dag",
             column_name="last_updated",
             type_=mysql.TIMESTAMP(fsp=6),
+            nullable=False,
```
Random comment: WHy don't we use DATETIME columns? https://dev.mysql.com/doc/refman/8.0/en/datetime.html

kaxil replied:
It didn't seem to keep time in UTC or support Time Zones. So kept it in-line with other tables 

https://github.com/apache/airflow/blob/a7a39003f97451347836acbc205d4ee2c35cbfd6/airflow/migrations/versions/0e2a74e0fc9f_add_time_zone_awareness.py#L54-L57

ashb replied:
Probably best to keep them all the same for now, but given our SQLA layer I don't think we _can_ write anything other than UTC in to the DB. 

But also I don't really care _what_ we do on Mysql so my caring is low.

ashb commented on airflow/api/common/experimental/delete_dag.py:
```diff
@@ -45,6 +49,11 @@ def delete_dag(dag_id: str, keep_records_in_log: bool = True, session=None) -> i
+    # Scheduler removes DAGs without files from serialized_dag table every dag_dir_list_interval.
+    # There may be a lag, so explicitly removes serialized DAG here.
+    if DAGCACHED_ENABLED and SerializedDagModel.has_dag(dag_id):
+        SerializedDagModel.remove_dag(dag_id)
```
```suggestion
        SerializedDagModel.remove_dag(dag_id, session=session)
```

May need to extend that method to take a session argument, but this way it all happens inside the same DB txn.

kaxil replied:
Updated in https://github.com/apache/airflow/pull/5743/commits/b814f8dfd9448ee3ceef2722c7f0291d8a680700

kaxil committed:
Add test to validate if task.subdag is None if operator is not SubDagOperator

Fix CI by filtering DAG list and removing SubDAGs

Figured out that the issue was because the it tries to remove SubDAG from DB which does not exist as a separate row and hence fails.

kaxil commented on tests/models/test_serialized_dag.py:
```diff
@@ -0,0 +1,106 @@
+    def test_remove_dags(self):
+        """DAGs can be removed from database."""
+        example_dags_list = list(self._write_example_dags().values())
+        print("1", example_dags_list)
```
Temporary - to remove

kaxil commented on tests/models/test_serialized_dag.py:
```diff
@@ -0,0 +1,106 @@
+        # Tests removing by file path.
+        dag_removed_by_file = example_dags_list[1]
+        example_dag_files = list([dag.full_filepath for dag in example_dags_list])
+        print("2", dag_removed_by_file)
```
Temporary - to remove

kaxil commented on tests/models/test_serialized_dag.py:
```diff
@@ -0,0 +1,106 @@
+        dag_removed_by_file = example_dags_list[1]
+        example_dag_files = list([dag.full_filepath for dag in example_dags_list])
+        print("2", dag_removed_by_file)
+        print("3", example_dag_files)
```
Temporary - to remove

kaxil commented on tests/models/test_serialized_dag.py:
```diff
@@ -0,0 +1,106 @@
+        print("2", dag_removed_by_file)
+        print("3", example_dag_files)
+        example_dag_files.remove(dag_removed_by_file.full_filepath)
+        print("4", example_dag_files)
```
Temporary - to remove

kaxil committed:
Use session whereever available

kaxil committed:
Fix operator displying as SerializedBaseOperator on UI

kaxil committed:
Add warning when dag_id does not match

kaxil committed:
Validate Json String before writing to DB

ashb commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,173 @@
+log = LoggingMixin().log
+
+
+class SerializedDagModel(Base):
```
```suggestion
class SerializedDag(Base):
```

And then call this `SerialzedDagModel` at import time if we need to distinguish this.

kaxil replied:
The problem is it will collide a lot in the same file & so become confusing:

https://github.com/apache/airflow/blob/e2570ca0002394af9f59703a2b29ad21642c546b/airflow/models/serialized_dag.py#L73

https://github.com/apache/airflow/blob/e2570ca0002394af9f59703a2b29ad21642c546b/airflow/models/serialized_dag.py#L34-L37`

I can obviously add `as` on those imports to change but not sure if it is worth it, WDYT?

ashb committed:
Store JSON schema as static, package-data JSON file

- Since it is now in a single document we can use internal schema
references to reuse common chunks

- Don't store dag_id against for serialized tasks

  We never store serialized operators apart from their dags, so we don't
  need to include the dag_id in the structure again. This change means
  we need to re-associate the task and the dag after inflating the dag,
  but does also mean we don't need to pass `visited_dags` around all over
  the place

- Always serialize template_fields as an list, never an array

  Although many operator classes set these as tuples instead of lists,
  that is a distinction that is not important to us here, and makes the
  schema more complex.

kaxil committed:
Remove unused fields from DAGs & move tests file

kaxil committed:
Update Serialisation tests after refactor

kaxil committed:
Update tree view to use dag from dagbag instead of creating new dagbag

kaxil committed:
Make Dag Schema Stricter: No Additional Properties allowed

kaxil committed:
Rename the dagcached config& dagcached_min_update_interval setting

kaxil committed:
Remove inlets & outlets from SerializedBaseOperator

Commit:
improve fileloc hashing in DAG persistence

kaxil committed:
Add documentation

BasPH commented on airflow/dag/serialization/schema.json:
```diff
@@ -0,0 +1,233 @@
+        },
+        "__var": {
+          "type": "number",
+          "exclusiveMinimum": 0
```
Just tried out this PR and hit an issue with `airflow db reset`: `jsonschema.exceptions.ValidationError: 0.0 is less than or equal to the minimum of 0` as a result of `timedelta(seconds=0)` in /airflow/tests/dags/test_retry_handling_job.py.

```suggestion
          "minimum": 0
```

kaxil committed:
Allow timedelta(seconds=0) in Serialized DAGs

kaxil committed:
Fix issue with different class for Pendulum Timezones

ashb commented on airflow/dag/serialization/serialization.py:
```diff
@@ -148,7 +148,7 @@ def _serialize(cls, var):  # pylint: disable=too-many-return-statements
             elif isinstance(var, datetime.timedelta):
                 return cls._encode(var.total_seconds(), type_=DAT.TIMEDELTA)
-            elif isinstance(var, pendulum.tz.Timezone):
+            elif isinstance(var, (pendulum.tz.Timezone, type(pendulum.UTC))):
```
This looks suspect - what about using `datetime.tzinfo` instead, which both pendulum.tz.Timezone and pendulum.tz.timezone_info.TimezoneInfo inherit from.

kaxil replied:
Thought about using that y'day but found an issue that Timezone and Timezoneinfo contains `name` property that we in the line below. datetime.tzinfo does not contain that

ashb replied:
Ah.

How about `pendulum.tz.timezone_info.TimezoneInfo` instead then - also `pendulum.UTC` is an instance, not the class itself. (The class is `pendulum.tz.timezone_info._UTC`)

kaxil replied:
I did have this (`pendulum.tz.timezone_info.TimezoneInfo`)  first, however IDE complained (and so I guess mypy might have complained too - I guess not sure though): 

![image](https://user-images.githubusercontent.com/8811558/64258521-50ea9f00-cf1f-11e9-9919-92bef96ac908.png)

And so I changed it to ` type(pendulum.UTC)` which checks the type of the instance created.

Happy to try `pendulum.tz.timezone_info.TimezoneInfo` and see if mypy passes/fails

ashb commented on airflow/dag/serialization/serialization.py:
```diff
@@ -176,43 +168,35 @@ def _serialize(cls, var, visited_dags):  # pylint: disable=too-many-return-state
             elif type_ == DAT.TIMEDELTA:
                 return datetime.timedelta(seconds=var)
             elif type_ == DAT.TIMEZONE:
                 return pendulum.tz.timezone(name=var)
```
According to the docs we should do this as

```suggestion
                 return pendulum.timezone(var)
```

Probably a bit safer to follow the docs.

kaxil replied:
agree, will update

kaxil committed:
Update timezone class

kaxil committed:
Parse date from timestamp instead of string

kaxil committed:
Do not serialize dates in tasks if they have matching date in DAG

kaxil committed:
Change type of data column to JSON & Add metric for dagbag size

kaxil committed:
Code Cleanup for JSON columns

kaxil committed:
Test code to allow old mysql & sqlite versions

kaxil committed:
Test to see if JSON column is available

kaxil committed:
Add Debug info

ashb committed:
Trial reducing size of SerializedDAGs

Don't store defaults, and remove many (but not all) of the type
annotations. in the JSON). Combined this reduces the size of our large
test DAGs to 40% of what they were

(cherry picked from commit a590ed3253f1f41e46a30df966cf17d5dae989f3)

ashb committed:
Add specific test for schedule_interval serialization

(cherry picked from commit 073464a17ab1b997b50994df7513b57e99f570c5)

ashb committed:
Support dateutil.relativedelta in SerializedDAGs

This was a valid type for schedule_interval already, so we should
continue supporting it

(cherry picked from commit ec9d705f1a90790bdcb099196269c77d3cc3d53c)

kaxil committed:
Cleanup

kaxil committed:
Fix imports for iSort

kaxil committed:
Delete non-existent Dags

kaxil committed:
Remove comment

Commit:
fix bugs that date/time/IntEnum are not supported in serialization.

kaxil committed:
Deactivate DAGs instead of deleting if their DAG file is deleted

kaxil committed:
Change maxDiff to None

kaxil committed:
Fix CI

kaxil committed:
Just-in-time loading of DagBag in webserver

To save start-up time (and memory) this changes the DabBag to not be
populated by the webserver on start up - and when a specific dag is
asked for it will be loaded on-demand from the SerializedDAG table.

Co-Authored-By: Ash Berlin-Taylor <ash_github@firemirror.com>

kaxil committed:
Add default args to decorated_fields

kaxil committed:
Add support for OperatorLinks

ExtraOperatorLinks are supported if Plugins are registered for them

ashb commented on airflow/plugins_manager.py:
```diff
@@ -57,8 +57,21 @@ class AirflowPlugin:
+    # Operator extra links have to be defined in plugins to work.
+    # These extra links will be available on the task page in form of
+    # buttons.
+    serialized_operator_extra_links = []  # type: List[BaseOperatorLink]
```
I would just call this `operator_extra_links` -- that way we don't have to rename it if/when we make serialized the only way of working.

kaxil replied:
Updated

ashb commented on docs/howto/define_extra_link.rst:
```diff
@@ -56,3 +56,27 @@ The following code shows how to add extra links to an operator:
+  # Defining the plugin class
+  class AirflowExtraLinkPlugin(AirflowPlugin):
+      name = "extra_link_plugin"
+      serialized_operator_extra_links = [GoogleLink(), ]
```
It's slightly odd that this is an instance here, rather than just a class.

ashb commented on tests/serialization/test_dag_serialization.py:
```diff
@@ -67,14 +69,49 @@
+                "ui_color": "#fff",
+                "ui_fgcolor": "#000",
+                "template_fields": [],
+                "extra_links": ["github"],
```
I'm not sure the Link name is unique enough -- Given the `link.name` attribute is what is displayed in the UI it is entirely reasonable for two different link classes to define a link with a name of "Logs" or "View in AWS" but that have different implementations.

I think `<plugin_name>.<LinkClass>` would be better, so `test_plugin.GithubLink`.

(To make this work we'd likely need to change a bit how we store plugins in airflow.plugins_manager)

kaxil replied:
I have updated it with a different approach: https://github.com/coufon/airflow/compare/d7bc655..39b5b7c?expand=1

Also created a separate PR for it https://github.com/apache/airflow/pull/6302

mik-laj commented on airflow/dag/serialization/schema.json:
```diff
@@ -0,0 +1,196 @@
+        "do_xcom_push": { "type": "boolean" },
+        "ui_color": { "$ref": "#/definitions/color" },
+        "ui_fgcolor": { "$ref": "#/definitions/color" },
+        "template_fields": {
```
Why is this not a `stringset` ? The values must be unique and the order does not matter.

ashb replied:
`stringset` isn't a valid type in JSON Schema that I can see...?

mik-laj replied:
This type has been previously defined.
https://github.com/apache/airflow/pull/5743/files#diff-c5a9d813838f2de6a2ffef12f84df3ddR76-R88

ashb replied:
Oh duh, i forgot about that :)

Mostly because the extra "type deocration" isn't needed anymore and we've removed it in cases where it is un-ambigious to the deserializer. (For the huge dags  we've been testing with the parsing time starts to have a non-trivial impact on performance). Since `stringset` isn't used anymore we should probably remove that definition instead.

kaxil replied:
Removed

mik-laj commented on airflow/dag/serialization/schema.json:
```diff
@@ -0,0 +1,196 @@
+        "_downstream_task_ids": {
+          "type": "array",
+          "items": { "type": "string" }
+ }
```
```suggestion
        }
```

kaxil replied:
Updated

mik-laj commented on airflow/dag/serialization/serialization.py:
```diff
@@ -53,81 +50,81 @@ class Serialization:
+        if isinstance(serialized_obj, dict):
+            cls._json_schema.validate(serialized_obj)
+        elif isinstance(serialized_obj, str):
+            cls._json_schema.validate(json.loads(serialized_obj))
```
```suggestion
            cls._json_schema.validate(json.loads(serialized_obj))
       else:
           raise TypeError("Invalid type: Only dict and str are supported.")
```
It is better not to return None in an undefined state.

kaxil replied:
Updated

ashb commented on tests/models/test_serialized_dag.py:
```diff
@@ -0,0 +1,104 @@
+from airflow.utils import db
+
+
+# FIXME: it is defined in tests/dags/test_dag_serialization.py as well.
```
We've removed this file in this PR

kaxil committed:
Cleanup

mik-laj commented on airflow/models/baseoperator.py:
```diff
@@ -1021,8 +1039,8 @@ def xcom_pull(
         """@property: extra links for the task. """
-        return list(set(self.operator_extra_link_dict.keys())
-                    .union(self.global_operator_extra_link_dict.keys()))
+        return sorted(list(set(self.operator_extra_link_dict.keys())
```
```suggestion
        return sorted(set(self.operator_extra_link_dict.keys())
```

ashb replied:
We might not need the set either - dicts can't have duplicate keys

```suggestion
        return sorted(self.operator_extra_link_dict.keys())
```

kaxil replied:
I am going to remove all the Extra link commits and changes to baseoperator once https://github.com/apache/airflow/pull/6302 is merged.

mik-laj commented on docs/howto/enable-dag-serialization.rst:
```diff
@@ -0,0 +1,109 @@
+
+
+
+Enable DAG Serialization
```
I think that you should first describe the feature and then describe how to enable it.

kaxil replied:
Updated.

mik-laj commented on docs/howto/enable-dag-serialization.rst:
```diff
@@ -0,0 +1,109 @@
+
+.. image:: ../img/dag_serialization.png
+
+As shown in the image above in Vanilla Airflow, the Webserver and the Scheduler both
```
```suggestion
In Vanilla Airflow, the Webserver and the Scheduler both
```

The term "Vanilia Airflow" seems strange to me. Are you sure you need to enter this terms in the documentation? I would prefer "Airflow with DAG persistence in DB" and "Airflow without DAG persistence in DB".

kaxil replied:
Agree. Updated

mik-laj commented on docs/howto/enable-dag-serialization.rst:
```diff
@@ -0,0 +1,109 @@
+With **DAG Serialization** we aim to decouple the webserver from DAG parsing
+which would make the Webserver very light-weight.
+
+As shown in the image above, when using the dag_serilization feature,
```
```suggestion
As shown in the image above, when using the this feature,
```

kaxil replied:
Updated

mik-laj commented on docs/howto/enable-dag-serialization.rst:
```diff
@@ -0,0 +1,109 @@
+*   :doc:`Extra Operator Links <define_extra_link>` would not work out of
+    the box. They need to be defined in Airflow Plugin.
+
+    **Existing Airflow Operators**:
```
I don't understand the relationship of this text to DAG serialization

milton0825 replied:
I think @kaxil is referring to existing operators with `operator extra links`

kaxil replied:
Yes, I was. Although now that we already have docs for adding operator using plugin merged, I will remove it from here

mik-laj commented on docs/howto/enable-dag-serialization.rst:
```diff
@@ -0,0 +1,109 @@
+
+    .. code-block:: python
+
+        from airflow.plugins_manager import AirflowPlugin
```
I'm not a fan of posting python code in rst files. We use literalinclude for GCP guides. This allows us to automatically test code snippets from the documentation

mik-laj commented on docs/howto/enable-dag-serialization.rst:
```diff
@@ -0,0 +1,109 @@
+            def get_link(self, operator, dttm):
+                ti = TaskInstance(task=operator, execution_date=dttm)
+                job_id = ti.xcom_pull(task_ids=operator.task_id, key='job_id')
+                return BIGQUERY_JOB_DETAILS_LINK_FMT.format(job_id=job_id) if job_id else ''
```
This code is not 100% valid. It has been updated. https://github.com/apache/airflow/pull/5906/files

kaxil replied:
Yes removed this code from this file.

Also, this OperatorLinks won't work with Serialized DAGs as described in  https://github.com/apache/airflow/pull/5743#discussion_r337801737

kaxil committed:
Move serialization directory out of dags folder

milton0825 commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,214 @@
+        :param min_update_interval: minimal interval in seconds to update serialized DAG
+        :param session: ORM Session
+        """
+        log.debug("Writing DAG: %s to the DB", dag)
```
```suggestion
        log.debug("Writing DAG: %s to the DB", dag.dag_id)
```

kaxil replied:
done

milton0825 commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,214 @@
+        """The DAG deserialized from the ``data`` column"""
+        from airflow.serialization import SerializedDAG  # noqa # pylint: disable=redefined-outer-name
+
+        if isinstance(self.data, dict):
```
Why would `self.data` have two formats here?

ashb replied:
This probably isn't needed, but might be for cases where SQLAlchemy doesn't handle the JSON deserialize for us.

We should remove this if we can though, yes.

kaxil replied:
Yes, older versions of Mysql don't support json type and sqlalchemy returns string. Best example is our CI images with MySQL

mik-laj replied:
Do you have information about which MySQL version does not support this?  Maybe it is worth abandoning the support for some versions to be able to go further in the future e.g. with JSON index etc.

kaxil replied:
>MySQL as of version 5.7 (MariaDB as of the 10.2 series does not)

https://docs.sqlalchemy.org/en/13/core/type_basics.html#sqlalchemy.types.JSON

milton0825 commented on docs/howto/enable-dag-serialization.rst:
```diff
@@ -0,0 +1,109 @@
+
+    **Existing Airflow Operators**:
+    To make extra operator links work with existing operators like BigQuery, copy all
+    the classes that are defined in ``operator_extra_links`` property.
```
Does it mean `operator_extra_links` can only be registered through plugin in the future?

ashb replied:
If dag serialization becomes the default yes. Well, either bundled in Airflow, or registered via a plugin.

@kaxil We should check that we serialize the "bundled" GCP links correctly (they should work as the webserver will have the code for this)

milton0825 replied:
Can you clarify a bit about the "bundled" GCP links? Are you talking about the operator extra links registered as a `@property` in the operator like:
https://github.com/apache/airflow/blob/master/airflow/gcp/operators/bigquery.py#L460-L470
?

kaxil replied:
Yes, unfortunately in the current implementation you need to add operator links via plugin for in-built operators too.

Let me try and see if I can remove the need to register links for in-built operators links

kaxil replied:
Looks like the currently defined OperatorLinks for inbuilt operators are dependent on the Instance of operator class.

Example the following is Qubole Operator:

```
class QDSLink(BaseOperatorLink):
    """Link to QDS"""
    name = 'Go to QDS'

    def get_link(self, operator, dttm):
        return operator.get_hook().get_extra_links(operator, dttm)
```
which depends on `conn = BaseHook.get_connection(operator.kwargs['qubole_conn_id'])` line. We don't store `qubole_conn_id` as the task property in our serialized DAG.

and the BigqueryOperator is:

```
@property
    def operator_extra_links(self):
        """
        Return operator extra lxinks
        """
        if isinstance(self.sql, str):
            return (
                BigQueryConsoleLink(),
            )
        return (
            BigQueryConsoleIndexableLink(i) for i, _ in enumerate(self.sql)
        )
```
this one is dependent on `self.sql` property !! which the Serialized Operator won't have access to as we don't store that info in serialized DAG.

cc @ashb  @milton0825

ashb replied:
Hmmmm. Tricky. And even if we did store the attributes (which is doable by setting `_serialized_fields` on the operator sub-classes) these links might not work for Quoble: `operator` will be SerializedBaseOperator, not a `QuobleOperator`.

(BQ would work well enough as the SQL would be of the right length, even if it was templated so the value stored in s10n is not the final value)

kaxil replied:
Exactly. Hence I feel to let the user create their own OperatorLink via plugin !

And storing properties like `self.sql` will increase the serialized blob size as it is a templated field it can take in a huge SQL Query.

milton0825 replied:
Should we create a pre-existing plugin and register the existing links there so that users can re-use them easily?

ashb replied:
I think for the pre-existing ones that do work we should do what is neccsary to registry them which since it is "internal" doesn't need to be a full plugin. Something like

```
import airflow.plugins_manager
airflow.plugins_manager.operator_extra_links.append(BigQueryConsoleLink)
```

sort of thing.

BasPH commented on airflow/models/serialized_dag.py:
```diff
@@ -0,0 +1,214 @@
+                     (timezone.utcnow() - timedelta(seconds=min_update_interval)) < cls.last_updated))
+            ).scalar():
+                return
+        session.merge(cls(dag))
```
Makes more sense to move the `log.debug("Writing DAG: %s to the DB", dag)` line here, because the interval check is done before writing the DAG.

kaxil replied:
👍

BasPH reviewed (commented):
Besides one tiny nit and the comments made by others, looks good to me!

Commit:
Fix docs building

kaxil committed:
Move the debug line before merging session

kaxil committed:
Move Dag Serialization doc from howto to root folder

KevinYang21 reviewed (approved):
LGTM together with a green CI
